File: 2. Structural Design Patterns/Flyweight/Javascript/FlyweightExample.js
// Flyweight Pattern Example in JavaScript
// Forest: share TreeType (intrinsic) among many Tree instances (extrinsic position)

class TreeType {
  constructor(name, color, texture){ this.name = name; this.color = color; this.texture = texture; }
  draw(x, y){ console.log(`Draw ${this.name} [${this.color},${this.texture}] at (${x},${y})`); }
}

class TreeFactory {
  static cache = new Map();
  static getTreeType(name, color, texture){
    const key = `${name}-${color}-${texture}`;
    if(!this.cache.has(key)) this.cache.set(key, new TreeType(name, color, texture));
    return this.cache.get(key);
  }
}

class Tree {
  constructor(x, y, type){ this.x = x; this.y = y; this.type = type; }
  draw(){ this.type.draw(this.x, this.y); }
}

class Forest {
  constructor(){ this.trees = []; }
  plant(x, y, name, color, texture){ const type = TreeFactory.getTreeType(name, color, texture); this.trees.push(new Tree(x,y,type)); }
  draw(){ for(const t of this.trees) t.draw(); }
}

(function main(){
  const forest = new Forest();
  for(let i=0;i<5;i++) forest.plant(i, i*2, 'Oak', 'Green', 'Rough');
  for(let i=0;i<3;i++) forest.plant(i+10, i*3, 'Pine', 'DarkGreen', 'Smooth');
  forest.draw();
})();

File: 2. Structural Design Patterns/Flyweight/Kotlin/FlyweightExample.kt
// Flyweight Pattern Example in Kotlin
// Forest: shares TreeType (intrinsic) among many Tree instances (extrinsic position)

class TreeType(private val name: String, private val color: String, private val texture: String) {
    fun draw(x: Int, y: Int) = println("Draw $name [$color,$texture] at ($x,$y)")
}

object TreeFactory {
    private val cache = mutableMapOf<String, TreeType>()
    fun getTreeType(name: String, color: String, texture: String): TreeType {
        val key = "$name-$color-$texture"
        return cache.getOrPut(key) { TreeType(name, color, texture) }
    }
}

class Tree(private val x: Int, private val y: Int, private val type: TreeType) {
    fun draw() = type.draw(x, y)
}

class Forest {
    private val trees = mutableListOf<Tree>()
    fun plant(x: Int, y: Int, name: String, color: String, texture: String) {
        val t = TreeFactory.getTreeType(name, color, texture)
        trees.add(Tree(x, y, t))
    }
    fun draw() = trees.forEach { it.draw() }
}

fun main() {
    val forest = Forest()
    repeat(5) { i -> forest.plant(i, i * 2, "Oak", "Green", "Rough") }
    repeat(3) { i -> forest.plant(i + 10, i * 3, "Pine", "DarkGreen", "Smooth") }
    forest.draw()
}
